### PR TITLE
Ensure images resources with long querystrings can be read

### LINKED ISSUE #3641: "[Errno 36] File name too long:" when using imread on remote resource with long querystring
## Description
When using skimage.io.imread with a remote resource, a long query string on the remote resource will cause a failure to read the remote resource, because the temporary file cannot be created.

e.g. 

The following works fine
```
>>> im = imread('https://c1.staticflickr.com/9/8370/8429454143_1066b73c04_o.jpg?{}'.format(''.join(['s' for i in range(100)])))

```

while the one below fails

```
>>> im = imread('https://c1.staticflickr.com/9/8370/8429454143_1066b73c04_o.jpg?{}'.format(''.join(['s' for i in range(300)])))
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/skimage/io/util.py", line 28, in file_or_url_context
    with tempfile.NamedTemporaryFile(delete=False, suffix=ext) as f:
  File "/usr/lib/python3.5/tempfile.py", line 688, in NamedTemporaryFile
    (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/usr/lib/python3.5/tempfile.py", line 399, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
OSError: [Errno 36] File name too long: '/tmp/tmpmfnujlq6.jpg?ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/skimage/io/_io.py", line 61, in imread
    with file_or_url_context(fname) as fname:
  File "/usr/lib/python3.5/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.5/dist-packages/skimage/io/util.py", line 34, in file_or_url_context
    os.remove(f.name)
UnboundLocalError: local variable 'f' referenced before assignment

```

## Way to reproduce
[If reporting a bug, please include the following important information:]
- [x] Code example
- [x] Operating system and version
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.5 LTS"
```
- [x] Python version: Python 3.5.2
- [x] scikit-image version (run `skimage.__version__`): skimage.__version__
'0.14.1'

## Proposal

https://github.com/scikit-image/scikit-image/blob/d24634d86e4f90fe96377209d66ed114b9b601e4/skimage/io/util.py#L22

The querystring should be excluded from the temporary file name extension

### PR DESCRIPTION
**Pull request by jo-tham**
## Description
Ensure that querystrings on urls are not included in the extension of the temporary file when reading a remote image. This avoids an Exception exceeding OS character limit due to querystring

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] Unit tests

NB: marked as complete as they are not applicable:
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark

[For detailed information on these and other aspects see [scikit-image contribution guidelines](https://scikit-image.org/docs/dev/contribute.html)]

## References
Fixes #3641

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
Did anybody else know about monkeypatch???/

File: skimage/io/tests/test_io.py [hunk 1]
```diff
@@ -26,3 +26,31 @@ def test_imread_url():
...
+                        b'\x10\x11\x10\n\x0c\x12\x13\x12\x10\x13\x0f\x10\x10'
+                        b'\x10\xff\xc0\x00\x0b\x08\x00\x01\x00\x01\x01\x01\x11'
+                        b'\x00\xff\xc4\x00\x14\x00\x01\x00\x00\x00\x00\x00\x00'
+                        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\t\xff\xc4\x00'
+                        b'\x14\x10\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+                        b'\x00\x00\x00\x00\x00\x00\xff\xda\x00\x08\x01\x01\x00'
+                        b'\x00?\x00*\x9f\xff\xd9')
+        return MockOpen
+
+    monkeypatch.setattr('skimage.io.util.urlopen', mockopen)
```
[the remark this answers is not in the document]
Comment (jo-tham) [reply skimage/io/tests/test_io.py]: named importation

``` python
from urllib.request import urlopen
```
creates a new name for the object, so mocking `urllib.request.urlopen` does not affect it.

however, the following would work. let me know if it is preferred

``` python
import urllib
...
            with tempfile.NamedTemporaryFile(delete=False, suffix=ext) as f:
                u = urllib.request.urlopen(resource_name)
```

[the remark this answers is not in the document]
Comment (jo-tham) [reply skimage/io/tests/test_io.py]: it's a bit annoying, but i guess an intrinsic side effect to named importation that mocking the original name does not work

[the remark this answers is not in the document]
Comment (jo-tham) [reply skimage/io/tests/test_io.py]: hard to find an explanation about this, there's this one, but it's python2. i don't see a py3 equivalent

https://docs.python.org/2.0/ref/import.html

The from form does not bind the module name: it goes through the list of identifiers, looks each one of them up in the module found in step (1), and binds the name in the local namespace to the object thus found. As with the first form of import, an alternate local name can be supplied by specifying "as localname". If a name is not found, ImportError is raised. If the list of identifiers is replaced by a star ("*"), all names defined in the module are bound, except those beginning with an underscore ("_").

**Comment by jo-tham** [comment]
Hi @stefanv & team, I think this is ready. Let me know if there's any changes required

**Comment by stefanv** [comment]
Thanks, that looks great!

**Comment by jo-tham** [comment]
No problem; thank you for the review and general contributions to skimage!

**Comment by hmaarrfk** [comment]
I have a general question about the approach:

is there any reason why we aren't just creating an image with a random filename???? it seems a little silly to be using the base image's name for a temporary file.

**Comment by jo-tham** [comment]
good question @hmaarrfk; this crossed my mind at some point ... I think my finding was that the file base name is temporary, but the extension of the original file was required because it is used (by an external depdency) to determine the probably encoding

File: skimage/io/util.py [hunk 2]
```diff
@@ -19,10 +19,11 @@ def is_url(filename):
 def file_or_url_context(resource_name):
     """Yield name of file from the given resource (i.e. file or url)."""
     if is_url(resource_name):
-        _, ext = os.path.splitext(resource_name)
+        url_components = urllib.parse.urlparse(resource_name)
+        _, ext = os.path.splitext(url_components.path)
         try:
             with tempfile.NamedTemporaryFile(delete=False, suffix=ext) as f:
```
Comment (hmaarrfk) [inline skimage/io/util.py]: I think this does create a random named temprory file.

It was just taking the whole `.jpeg?laskdjflkasdjflkjasdflkjsad;fj` as the extension

**Comment by hmaarrfk** [comment]
@jo-tham I think I finally get it. This whole context manager in a context manager makes my head hurt. But I think it was getting confused by the ?

**Comment by jni** [comment]
Thanks @jo-tham!

**Pull request merged**